### PR TITLE
[HttpFoundation] add sessions

### PR DESCRIPTION
**Pull request by alekitto**
Adds sessions, cache pools configuration and filesystem cache support

### CODE REVIEW

**Comment by massimilianobraglia** [comment]
Great work! I'll review as soon as possible!!

File: src/Component/Cache/src/CacheItemInterface.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ class CacheItemInterface {
     /**
      * Sets the value represented by this cache item.
      *
-     * The $value argument may be any item that can be serialized by PHP,
+     * The $value argument may be any item that can be serialized,
```
Comment (massimilianobraglia) [inline src/Component/Cache/src/CacheItemInterface.js]: `The $value argument may be any item that can be serialized,`
Also the `$` should be removed

File: src/Component/HttpFoundation/src/Response.js [hunk 2]
```diff
@@ -256,6 +256,269 @@ class Response {
...
+
+        return this;
+    }
+
+    /**
+     * Returns the number of seconds after the time specified in the response's Date
+     * header when the response should no longer be considered fresh.
+     *
+     * First, it checks for a s-maxage directive, then a max-age directive, and then it falls
+     * back on an expires header. It returns null when no maximum age can be established.
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Response.js]: It returns ~~null~~ `undefined` when no maximum age can be established.

File: src/Component/HttpFoundation/src/Response.js [hunk 3]
```diff
@@ -256,6 +256,269 @@ class Response {
...
+        this.setPublic();
+        this.headers.addCacheControlDirective('s-maxage', value);
+
+        return this;
+    }
+
+    /**
+     * Returns the response's time-to-live in seconds.
+     *
+     * It returns null when no freshness information is present in the response.
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Response.js]: It returns ~~null~~ `undefined` when no freshness information is present in the response.

File: src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ * FlashBagInterface.
+ *
+ * @memberOf Jymfony.Component.HttpFoundation.Session.Flash
+ */
+class FlashBagInterface extends SessionBagInterface.definition {
+    /**
+     * Adds a flash message for type.
+     *
+     * @param {string} type
+     * @param {*} message
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js:13]: Should the message be a string?

File: src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+     * @param {string}       type
+     * @param {string|string[]} message
+     */
+    set(type, message) { }
+
+    /**
+     * Gets flash messages for a given type.
+     *
+     * @param {string} type Message category type
+     * @param {string[]} defaultValue Default value if $type does not exist
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js]: @param {string[]} *[defaultValue = []]* Default value if ~~$type~~ *type* does not exist

File: src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * @returns {Object.<string, string[]>}
+     */
+    peekAll() { }
+
+    /**
+     * Gets and clears flash from the stack.
+     *
+     * @param {string} type
+     * @param {string[]} defaultValue Default value if $type does not exist
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js]: Same as above

File: src/Component/HttpFoundation/src/Session/Session.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+/**
+ * @memberOf Jymfony.Component.HttpFoundation.Session
+ */
+class Session extends implementationOf(SessionInterface) {
+    /**
+     * Constructor.
+     *
+     * @param {Jymfony.Component.HttpFoundation.Session.Storage.SessionStorageInterface} storage
+     * @param {Jymfony.Component.HttpFoundation.Session.AttributeBagInterface} attributeBag
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/Session.js]: JSDocs... 😄

File: src/Component/HttpFoundation/src/Session/SessionInterface.js [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+ *
+ * @memberOf Jymfony.Component.HttpFoundation.Session
+ */
+class SessionInterface {
+    /**
+     * Starts the session storage.
+     *
+     * @returns {Promise<boolean>} True if session started
+     *
+     * @throws {RuntimeException} if session fails to start
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/SessionInterface.js:12]: Maybe we can think about the return value... It doesn't make any sense to return the `Promise<boolean>` on success if on failure a `RuntimeException` is thrown 😄

File: src/Component/HttpFoundation/src/Session/SessionInterface.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+     */
+    set name(name) { }
+
+    /**
+     * Invalidates the current session.
+     *
+     * Clears all session attributes and flashes and regenerates the
+     * session and deletes the old session from persistence.
+     *
+     * @param {int} [lifetime] Sets the cookie lifetime for the session cookie. 0 sets the cookie
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/SessionInterface.js]: `@param {int} [lifetime = 0]`

File: src/Component/HttpFoundation/src/Session/SessionInterface.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+     * @returns {Promise<boolean>} True if session invalidated, false if error
+     */
+    async invalidate(lifetime = 0) { }
+
+    /**
+     * Migrates the current session to a new session id while maintaining all
+     * session attributes.
+     *
+     * @param {boolean} destroy Whether to delete the old session or leave it to garbage collection
+     * @param {int} lifetime Sets the cookie lifetime for the session cookie. 0 sets the cookie
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/SessionInterface.js]: `@param {boolean} [destroy = false]`
`@param {int} [lifetime = 0]`

File: src/Component/HttpFoundation/src/Session/Storage/CacheSessionStorage.js [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        this._id = undefined;
+
+        /**
+         * Session name.
+         *
+         * @type {string}
+         *
+         * @private
+         */
+        this._name = 'JSSESSID';
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/Storage/CacheSessionStorage.js]: `JFSESSID`? 😎

File: src/Component/HttpFoundation/src/Session/Storage/SessionStorageInterface.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ * @memberOf Jymfony.Component.HttpFoundation.Session.Storage
+ */
+class SessionStorageInterface {
+    /**
+     * Starts the session.
+     *
+     * @returns {Promise<boolean>} True if started
+     *
+     * @throws {RuntimeException} if something goes wrong starting the session
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Session/Storage/SessionStorageInterface.js:12]: And again: it does not make any sense this return value 😅

File: src/Component/HttpServer/src/EventListener/AbstractSessionListener.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+             * can still operate on the open session. This prevents the overhead of restarting it.
+             * Listeners after closing the session can still work with the session as usual because
+             * Jymfony's session implementation starts the session on demand. So writing to it after
+             * it is saved will just restart it.
+             */
+            session.save();
+        }
+    }
+
+    static getSubscribedEvents() {
```
Comment (massimilianobraglia) [inline src/Component/HttpServer/src/EventListener/AbstractSessionListener.js:77]:
```
/** 
 * @inheritdoc
 */
```

File: src/Component/Cache/src/Adapter/FilesystemAdapter.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+const AbstractAdapter = Jymfony.Component.Cache.Adapter.AbstractAdapter;
+
+/**
+ * @memberOf Jymfony.Component.Cache.Adapter
+ */
+class FilesystemAdapter extends mix(AbstractAdapter) {
+    __construct(namespace = '', defaultLifetime = 0, directory = undefined) {
```
Comment (massimilianobraglia) [inline src/Component/Cache/src/Adapter/FilesystemAdapter.js:15]: missin' jsdocs

File: src/Component/HttpFoundation/src/Response.js [hunk 15]
```diff
@@ -256,6 +256,269 @@ class Response {
...
+    /**
+     * Sets the Expires HTTP header with a DateTime instance.
+     *
+     * Passing null as value will remove the header.
+     *
+     * @returns {Jymfony.Component.HttpFoundation.Response}
+     *
+     * @final
+     */
+    setExpires(date = undefined) {
```
Comment (massimilianobraglia) [inline src/Component/HttpFoundation/src/Response.js:394]: missing `@param` doc

**Review by massimilianobraglia — CHANGES_REQUESTED** [review]
This is a great contribution, good job @alekitto!

Just left a few comments plus the next one:
This is a general comment about the project, not this specific feature:

there are some points in the codebase like the following that there are not checks about possible keys of option objects. Should we use the OptionsResolver? Or, if we don't want a new dependency for the components, should be there anyway a precise exception thrown instead of an error?

```js
super.__construct(new FilesystemAdapter('', options.max_lifetime, options.save_path), options.max_lifetime);
```

File: src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js (hunk 4 above)
Comment (alekitto) [reply src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js:13]: Ideally `message` can be anything and should be converted to a string using `.toString()`

Comment (massimilianobraglia) [reply src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js:13]: Ok, then we should explicit it as a MUST for implementors. Anyway in this one message is documented as `string|string[]`
https://github.com/jymfony/jymfony/blob/681ebab381011d774e61c40c50375681e1930d10/src/Component/HttpFoundation/src/Session/Flash/FlashBagInterface.js#L21

**Pull request merged**